File: src/main/java/space/npstr/wolfia/domain/stats/StatsRepository.java
/*
 * Copyright (C) 2016-2020 the original author or authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package space.npstr.wolfia.domain.stats;

import io.prometheus.client.Summary;
import org.jooq.DSLContext;
import org.jooq.Record1;
import org.jooq.RecordMapper;
import org.jooq.impl.DSL;
import org.springframework.stereotype.Repository;
import space.npstr.wolfia.db.AsyncDbWrapper;
import space.npstr.wolfia.db.gen.tables.records.StatsActionRecord;
import space.npstr.wolfia.db.gen.tables.records.StatsPlayerRecord;
import space.npstr.wolfia.db.gen.tables.records.StatsTeamRecord;
import space.npstr.wolfia.game.definitions.Alignments;
import space.npstr.wolfia.system.metrics.MetricsRegistry;

import javax.annotation.CheckReturnValue;
import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletionStage;

import static org.jooq.impl.DSL.avg;
import static org.jooq.impl.DSL.count;
import static org.jooq.impl.DSL.defaultValue;
import static space.npstr.wolfia.db.gen.Tables.STATS_ACTION;
import static space.npstr.wolfia.db.gen.Tables.STATS_GAME;
import static space.npstr.wolfia.db.gen.Tables.STATS_PLAYER;
import static space.npstr.wolfia.db.gen.Tables.STATS_TEAM;

@Repository
public class StatsRepository {

    private final AsyncDbWrapper wrapper;

    public StatsRepository(AsyncDbWrapper wrapper) {
        this.wrapper = wrapper;
    }

    @CheckReturnValue
    public CompletionStage<BigDecimal> getAveragePlayerSize() {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getAveragePlayerSize");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(avg(STATS_GAME.PLAYER_SIZE))
                .from(STATS_GAME)
                .fetchOptional() // SQL AVG may return null for empty sets
                .map(Record1::component1)
                .orElse(BigDecimal.ZERO)
        ));
    }

    @CheckReturnValue
    public CompletionStage<BigDecimal> getAveragePlayerSizeInGuild(long guildId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getAveragePlayerSizeInGuild");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(avg(STATS_GAME.PLAYER_SIZE))
                .from(STATS_GAME)
                .where(STATS_GAME.GUILD_ID.eq(guildId))
                .fetchOptional() // SQL AVG may return null for empty sets
                .map(Record1::component1)
                .orElse(BigDecimal.ZERO)
        ));
    }

    @CheckReturnValue
    public CompletionStage<Set<Integer>> getDistinctPlayerSizes() {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getDistinctPlayerSizes");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .selectDistinct(STATS_GAME.PLAYER_SIZE)
                .from(STATS_GAME)
                .fetch()
                .intoSet(STATS_GAME.PLAYER_SIZE)
        ));
    }

    @CheckReturnValue
    public CompletionStage<Set<Integer>> getDistinctPlayerSizesInGuild(long guildId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getDistinctPlayerSizesInGuild");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .selectDistinct(STATS_GAME.PLAYER_SIZE)
                .from(STATS_GAME)
                .where(STATS_GAME.GUILD_ID.eq(guildId))
                .fetch()
                .intoSet(STATS_GAME.PLAYER_SIZE)
        ));
    }

    @CheckReturnValue
    public CompletionStage<Integer> countAlignmentWins(Alignments alignment) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("countAlignmentWins");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(count())
                .from(STATS_GAME)
                .innerJoin(STATS_TEAM)
                .on(STATS_TEAM.GAME_ID.eq(STATS_GAME.GAME_ID))
                .where(STATS_TEAM.IS_WINNER.isTrue())
                .and(STATS_TEAM.ALIGNMENT.eq(alignment.name()))
                .fetchOne()
                .component1()
        ));
    }

    @CheckReturnValue
    public CompletionStage<Integer> countAlignmentWinsInGuild(Alignments alignment, long guildId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("countAlignmentWinsInGuild");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(count())
                .from(STATS_GAME)
                .innerJoin(STATS_TEAM)
                .on(STATS_TEAM.GAME_ID.eq(STATS_GAME.GAME_ID))
                .where(STATS_TEAM.IS_WINNER.isTrue())
                .and(STATS_TEAM.ALIGNMENT.eq(alignment.name()))
                .and(STATS_GAME.GUILD_ID.eq(guildId))
                .fetchOne()
                .component1()
        ));
    }

    @CheckReturnValue
    public CompletionStage<Integer> countAlignmentWinsForPlayerSize(Alignments alignment, int playerSize) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("countAlignmentWinsForPlayerSize");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(count())
                .from(STATS_GAME)
                .innerJoin(STATS_TEAM)
                .on(STATS_TEAM.GAME_ID.eq(STATS_GAME.GAME_ID))
                .where(STATS_TEAM.IS_WINNER.isTrue())
                .and(STATS_GAME.PLAYER_SIZE.eq(playerSize))
                .and(STATS_TEAM.ALIGNMENT.eq(alignment.name()))
                .fetchOne()
                .component1()
        ));
    }

    @CheckReturnValue
    public CompletionStage<Integer> countAlignmentWinsForPlayerSizeInGuild(Alignments alignment, int playerSize, long guildId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("countAlignmentWinsForPlayerSizeInGuild");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(count())
                .from(STATS_GAME)
                .innerJoin(STATS_TEAM)
                .on(STATS_TEAM.GAME_ID.eq(STATS_GAME.GAME_ID))
                .where(STATS_TEAM.IS_WINNER.isTrue())
                .and(STATS_GAME.PLAYER_SIZE.eq(playerSize))
                .and(STATS_TEAM.ALIGNMENT.eq(alignment.name()))
                .and(STATS_GAME.GUILD_ID.eq(guildId))
                .fetchOne()
                .component1()
        ));
    }

    @CheckReturnValue
    public CompletionStage<List<GeneralUserStats>> getGeneralUserStats(long userId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getGeneralUserStats");
        return this.wrapper.jooq(dsl -> timer.time(() -> dsl
                .select(STATS_PLAYER.TOTAL_POSTLENGTH, STATS_PLAYER.TOTAL_POSTS, STATS_PLAYER.ALIGNMENT, STATS_TEAM.IS_WINNER)
                .from(STATS_PLAYER)
                .innerJoin(STATS_TEAM).on(STATS_PLAYER.TEAM_ID.eq(STATS_TEAM.TEAM_ID))
                .where(STATS_PLAYER.USER_ID.eq(userId))
                .fetchInto(ImmutableGeneralUserStats.class)
        ));
    }

    @CheckReturnValue
    public CompletionStage<List<String>> getUserShots(long userId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("getUserShots");
        return this.wrapper.jooq(dsl -> timer.time(() -> Arrays.asList(dsl
                .select(STATS_PLAYER.ALIGNMENT)
                .from(STATS_ACTION)
                .innerJoin(STATS_PLAYER).on(STATS_PLAYER.USER_ID.eq(STATS_ACTION.TARGET))
                .innerJoin(STATS_TEAM).on(STATS_TEAM.TEAM_ID.eq(STATS_PLAYER.TEAM_ID))
                .innerJoin(STATS_GAME).on(STATS_ACTION.GAME_ID.eq(STATS_GAME.GAME_ID).and(STATS_TEAM.GAME_ID.eq(STATS_GAME.GAME_ID)))
                .where(STATS_ACTION.ACTION_TYPE.eq("SHOOT").and(STATS_ACTION.ACTOR.eq(userId)))
                .fetch()
                .intoArray(STATS_PLAYER.ALIGNMENT)
        )));
    }

    @CheckReturnValue
    public CompletionStage<Optional<GameStats>> findGameStats(long gameId) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("findGameStats");
        return this.wrapper.jooq(dsl -> timer.time(() -> {
                    Optional<GameStats> gameOpt = dsl.selectFrom(STATS_GAME)
                            .where(STATS_GAME.GAME_ID.eq(gameId))
                            .fetchOptionalInto(GameStats.class);

                    if (gameOpt.isEmpty()) {
                        return gameOpt;
                    }

                    GameStats game = gameOpt.get();

                    List<TeamStats> teams = dsl.selectFrom(STATS_TEAM)
                            .where(STATS_TEAM.GAME_ID.eq(gameId))
                            .fetch(teamMapper(game));
                    game.setTeams(teams);

                    for (TeamStats teamStats : teams) {
                        List<PlayerStats> players = dsl.selectFrom(STATS_PLAYER)
                                .where(STATS_PLAYER.TEAM_ID.eq(teamStats.getTeamId().orElseThrow()))
                                .fetch(playerMapper(teamStats));
                        teamStats.setPlayers(players);
                    }

                    List<ActionStats> actions = dsl.selectFrom(STATS_ACTION)
                            .where(STATS_ACTION.GAME_ID.eq(gameId))
                            .fetch(actionMapper(game));
                    game.setActions(actions);

                    return Optional.of(game);
                }
        ));
    }

    @CheckReturnValue
    public CompletionStage<GameStats> insertGameStats(GameStats gameStats) {
        Summary.Child timer = MetricsRegistry.queryTime.labels("insertGameStats");
        return this.wrapper.jooq(dsl -> dsl.transactionResult(config -> timer.time(() -> {
                    DSLContext context = DSL.using(config);

                    long gameId = context
                            .insertInto(STATS_GAME)
                            .values(defaultValue(STATS_GAME.GAME_ID), gameStats.getChannelId(),
                                    gameStats.getChannelName(), gameStats.getEndTime(), gameStats.getGameMode().name(),
                                    gameStats.getGameType().name(), gameStats.getGuildId(), gameStats.getGuildName(),
                                    gameStats.getStartTime(), gameStats.getPlayerSize())
                            .returningResult(STATS_GAME.GAME_ID)
                            .fetchOne()
                            .component1();
                    gameStats.setGameId(gameId);

                    for (TeamStats teamStats : gameStats.getStartingTeams()) {
                        long teamId = context
                                .insertInto(STATS_TEAM)
                                .values(defaultValue(STATS_TEAM.TEAM_ID), teamStats.getAlignment(),
                                        teamStats.isWinner(), teamStats.getName(), gameId, teamStats.getTeamSize())
                                .returningResult(STATS_TEAM.TEAM_ID)
                                .fetchOne()
                                .component1();
                        teamStats.setTeamId(teamId);


                        for (PlayerStats playerStats : teamStats.getPlayers()) {
                            long playerId = context
                                    .insertInto(STATS_PLAYER)
                                    .values(defaultValue(STATS_PLAYER.PLAYER_ID), playerStats.getNickname(),
                                            playerStats.getRole().name(), playerStats.getTotalPostLength(),
                                            playerStats.getTotalPosts(), playerStats.getUserId(), teamId,
                                            playerStats.getAlignment().name())
                                    .returningResult(STATS_PLAYER.PLAYER_ID)
                                    .fetchOne()
                                    .component1();
                            playerStats.setPlayerId(playerId);
                        }
                    }

                    for (ActionStats actionStats : gameStats.getActions()) {
                        long actionId = context
                                .insertInto(STATS_ACTION)
                                .values(defaultValue(STATS_ACTION.ACTION_ID), actionStats.getActionType().name(),
                                        actionStats.getActor(), actionStats.getCycle(), actionStats.getOrder(),
                                        actionStats.getTarget(), actionStats.getTimeStampHappened(),
                                        actionStats.getTimeStampSubmitted(), gameId, actionStats.getPhase().name(),
                                        actionStats.getAdditionalInfo())
                                .returningResult(STATS_ACTION.ACTION_ID)
                                .fetchOne()
                                .component1();
                        actionStats.setActionId(actionId);
                    }

                    return gameStats;
                }
        )));
    }

    private RecordMapper<StatsTeamRecord, TeamStats> teamMapper(GameStats gameStats) {
        return record -> new TeamStats(record.getTeamId(), record.getAlignment(), record.getIsWinner(),
                record.getName(), gameStats, record.getTeamSize());
    }

    private RecordMapper<StatsPlayerRecord, PlayerStats> playerMapper(TeamStats teamStats) {
        return record -> new PlayerStats(record.getPlayerId(), record.getNickname(), record.getRole(),
                record.getTotalPostlength(), record.getTotalPosts(), record.getUserId(), teamStats, record.getAlignment());
    }

    private RecordMapper<StatsActionRecord, ActionStats> actionMapper(GameStats gameStats) {
        return record -> new ActionStats(record.getActionId(), record.getActionType(), record.getActor(),
                record.getCycle(), record.getSequence(), record.getTarget(), record.getHappened(),
                record.getSubmitted(), gameStats, record.getPhase(), record.getAdditionalInfo());
    }
}

File: src/main/java/space/npstr/wolfia/domain/stats/StatsRepository.kt
/*
 * Copyright (C) 2016-2025 the original author or authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package space.npstr.wolfia.domain.stats

import java.math.BigDecimal
import java.time.Instant
import org.jooq.DSLContext
import org.jooq.Record3
import org.jooq.Record8
import org.jooq.RecordMapper
import org.jooq.impl.DSL
import org.springframework.stereotype.Repository
import space.npstr.wolfia.db.gen.Tables
import space.npstr.wolfia.db.gen.tables.records.StatsActionRecord
import space.npstr.wolfia.db.gen.tables.records.StatsPlayerRecord
import space.npstr.wolfia.db.gen.tables.records.StatsTeamRecord
import space.npstr.wolfia.domain.privacy.PrivacyAction
import space.npstr.wolfia.domain.privacy.PrivacyGame
import space.npstr.wolfia.game.GameInfo
import space.npstr.wolfia.game.definitions.Actions
import space.npstr.wolfia.game.definitions.Alignments
import space.npstr.wolfia.game.definitions.Games
import space.npstr.wolfia.game.definitions.Phase
import space.npstr.wolfia.game.definitions.Roles
import space.npstr.wolfia.system.metrics.MetricsService
import space.npstr.wolfia.system.metrics.time

@Repository
class StatsRepository internal constructor(
	private val jooq: DSLContext,
	private val metricsService: MetricsService,
) {

	fun fetchAveragePlayerSize(): BigDecimal {
		return metricsService.queryTime("getAveragePlayerSize").time {
			jooq
				.select(DSL.avg(Tables.STATS_GAME.PLAYER_SIZE))
				.from(Tables.STATS_GAME)
				.fetchOne()?.component1() ?: BigDecimal.ZERO // SQL AVG may return null for empty sets
		}
	}

	fun fetchAveragePlayerSizeInGuild(guildId: Long): BigDecimal {
		return metricsService.queryTime("getAveragePlayerSizeInGuild").time {
			jooq
				.select(DSL.avg(Tables.STATS_GAME.PLAYER_SIZE))
				.from(Tables.STATS_GAME)
				.where(Tables.STATS_GAME.GUILD_ID.eq(guildId))
				.fetchOne()?.component1() ?: BigDecimal.ZERO // SQL AVG may return null for empty sets
		}
	}

	fun fetchDistinctPlayerSizes(): Set<Int> {
		return metricsService.queryTime("getDistinctPlayerSizes").time {
			jooq
				.selectDistinct(Tables.STATS_GAME.PLAYER_SIZE)
				.from(Tables.STATS_GAME)
				.fetch()
				.intoSet(Tables.STATS_GAME.PLAYER_SIZE)
		}
	}

	fun fetchDistinctPlayerSizesInGuild(guildId: Long): Set<Int> {
		return metricsService.queryTime("getDistinctPlayerSizesInGuild").time {
			jooq
				.selectDistinct(Tables.STATS_GAME.PLAYER_SIZE)
				.from(Tables.STATS_GAME)
				.where(Tables.STATS_GAME.GUILD_ID.eq(guildId))
				.fetch()
				.intoSet(Tables.STATS_GAME.PLAYER_SIZE)
		}
	}

	fun countAlignmentWins(alignment: Alignments): Int {
		return metricsService.queryTime("countAlignmentWins").time {
			jooq
				.select(DSL.count())
				.from(Tables.STATS_GAME)
				.innerJoin(Tables.STATS_TEAM)
				.on(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.where(Tables.STATS_TEAM.IS_WINNER.isTrue)
				.and(Tables.STATS_TEAM.ALIGNMENT.eq(alignment.name))
				.fetchSingle()
				.component1()

		}
	}

	fun countAlignmentWinsInGuild(alignment: Alignments, guildId: Long): Int {
		return metricsService.queryTime("countAlignmentWinsInGuild").time {
			jooq
				.select(DSL.count())
				.from(Tables.STATS_GAME)
				.innerJoin(Tables.STATS_TEAM)
				.on(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.where(Tables.STATS_TEAM.IS_WINNER.isTrue)
				.and(Tables.STATS_TEAM.ALIGNMENT.eq(alignment.name))
				.and(Tables.STATS_GAME.GUILD_ID.eq(guildId))
				.fetchSingle()
				.component1()
		}
	}

	fun countAlignmentWinsForPlayerSize(alignment: Alignments, playerSize: Int): Int {
		return metricsService.queryTime("countAlignmentWinsForPlayerSize").time {
			jooq
				.select(DSL.count())
				.from(Tables.STATS_GAME)
				.innerJoin(Tables.STATS_TEAM)
				.on(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.where(Tables.STATS_TEAM.IS_WINNER.isTrue)
				.and(Tables.STATS_GAME.PLAYER_SIZE.eq(playerSize))
				.and(Tables.STATS_TEAM.ALIGNMENT.eq(alignment.name))
				.fetchSingle()
				.component1()
		}
	}

	fun countAlignmentWinsForPlayerSizeInGuild(alignment: Alignments, playerSize: Int, guildId: Long): Int {
		return metricsService.queryTime("countAlignmentWinsForPlayerSizeInGuild").time {
			jooq
				.select(DSL.count())
				.from(Tables.STATS_GAME)
				.innerJoin(Tables.STATS_TEAM)
				.on(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.where(Tables.STATS_TEAM.IS_WINNER.isTrue)
				.and(Tables.STATS_GAME.PLAYER_SIZE.eq(playerSize))
				.and(Tables.STATS_TEAM.ALIGNMENT.eq(alignment.name))
				.and(Tables.STATS_GAME.GUILD_ID.eq(guildId))
				.fetchSingle()
				.component1()
		}
	}

	fun fetchGeneralUserStats(userId: Long): List<GeneralUserStats> {
		return metricsService.queryTime("getGeneralUserStats").time {
			jooq
				.select(
					Tables.STATS_PLAYER.TOTAL_POSTLENGTH,
					Tables.STATS_PLAYER.TOTAL_POSTS,
					Tables.STATS_PLAYER.ALIGNMENT,
					Tables.STATS_TEAM.IS_WINNER,
				)
				.from(Tables.STATS_PLAYER)
				.innerJoin(Tables.STATS_TEAM).on(Tables.STATS_PLAYER.TEAM_ID.eq(Tables.STATS_TEAM.TEAM_ID))
				.where(Tables.STATS_PLAYER.USER_ID.eq(userId))
				.fetch()
				.map {
					GeneralUserStats(
						it.get(Tables.STATS_PLAYER.TOTAL_POSTLENGTH),
						it.get(Tables.STATS_PLAYER.TOTAL_POSTS),
						Alignments.valueOf(it.get(Tables.STATS_PLAYER.ALIGNMENT)),
						it.get(Tables.STATS_TEAM.IS_WINNER),
					)
				}
		}
	}

	fun fetchUserShots(userId: Long): List<Alignments> {
		return metricsService.queryTime("getUserShots").time {
			jooq
				.select(Tables.STATS_PLAYER.ALIGNMENT)
				.from(Tables.STATS_ACTION)
				.innerJoin(Tables.STATS_PLAYER).on(Tables.STATS_PLAYER.USER_ID.eq(Tables.STATS_ACTION.TARGET))
				.innerJoin(Tables.STATS_TEAM).on(Tables.STATS_TEAM.TEAM_ID.eq(Tables.STATS_PLAYER.TEAM_ID))
				.innerJoin(Tables.STATS_GAME)
				.on(
					Tables.STATS_ACTION.GAME_ID.eq(Tables.STATS_GAME.GAME_ID)
						.and(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID)),
				)
				.where(Tables.STATS_ACTION.ACTION_TYPE.eq("SHOOT").and(Tables.STATS_ACTION.ACTOR.eq(userId)))
				.fetch()
				.intoArray(Tables.STATS_PLAYER.ALIGNMENT)
				.toList()
				.map { Alignments.valueOf(it) }
		}
	}

	fun findGameStats(gameId: Long): GameStats? {
		return metricsService.queryTime("findGameStats").time {
			val dsl = jooq
			val gameRecord = dsl.selectFrom(Tables.STATS_GAME)
				.where(Tables.STATS_GAME.GAME_ID.eq(gameId))
				.fetchOne()
				?: return@time null

			val teams = dsl.selectFrom(Tables.STATS_TEAM)
				.where(Tables.STATS_TEAM.GAME_ID.eq(gameId))
				.fetch()
				.map { teamRecord ->
					val players = dsl.selectFrom(Tables.STATS_PLAYER)
						.where(Tables.STATS_PLAYER.TEAM_ID.eq(teamRecord.teamId))
						.fetch(playerMapper())
					return@map mapTeam(teamRecord, players)
				}

			val actions = dsl.selectFrom(Tables.STATS_ACTION)
				.where(Tables.STATS_ACTION.GAME_ID.eq(gameId))
				.fetch(actionMapper())

			return@time GameStats(
				gameRecord.gameId,
				gameRecord.channelId,
				gameRecord.channelName,
				gameRecord.endTime,
				GameInfo.GameMode.valueOf(gameRecord.gameMode),
				Games.valueOf(gameRecord.gameType),
				gameRecord.guildId,
				gameRecord.guildName,
				gameRecord.startTime,
				gameRecord.playerSize,
				teams.toSet(),
				actions.toSet(),
			)
		}
	}

	private fun mapTeam(teamRecord: StatsTeamRecord, players: Collection<PlayerStats>): TeamStats {
		return TeamStats(
			teamRecord.teamId, Alignments.valueOf(teamRecord.alignment), teamRecord.isWinner,
			teamRecord.name, teamRecord.teamSize, players.toSet(),
		)
	}

	private fun playerMapper(): RecordMapper<StatsPlayerRecord, PlayerStats> {
		return RecordMapper { record: StatsPlayerRecord ->
			PlayerStats(
				record.playerId, record.nickname, Roles.valueOf(record.role),
				record.totalPostlength, record.totalPosts, record.userId, Alignments.valueOf(record.alignment),
			)
		}
	}

	private fun actionMapper(): RecordMapper<StatsActionRecord, ActionStats> {
		return RecordMapper { record: StatsActionRecord ->
			ActionStats(
				record.actionId, Actions.valueOf(record.actionType), record.actor,
				record.cycle, record.sequence, record.target, record.happened,
				record.submitted, Phase.valueOf(record.phase), record.additionalInfo,
			)
		}
	}

	fun insertGameStats(insertGameStats: InsertGameStats): GameStats {
		return metricsService.queryTime("insertGameStats").time {
			jooq.transactionResult { config ->
				val context = config.dsl()
				val gameId = context
					.insertInto(Tables.STATS_GAME)
					.values(
						DSL.defaultValue(Tables.STATS_GAME.GAME_ID), insertGameStats.channelId,
						insertGameStats.channelName, insertGameStats.endTime, insertGameStats.gameMode.name,
						insertGameStats.gameType.name, insertGameStats.guildId, insertGameStats.guildName,
						insertGameStats.startTime, insertGameStats.playerSize,
					)
					.returningResult(Tables.STATS_GAME.GAME_ID)
					.fetchSingle()
					.component1()


				val teams = insertGameStats.getStartingTeams().map mapTeam@{ teamStats ->
					val teamRecord = context
						.insertInto(Tables.STATS_TEAM)
						.values(
							DSL.defaultValue(Tables.STATS_TEAM.TEAM_ID), teamStats.alignment,
							teamStats.isWinner, teamStats.name, gameId, teamStats.teamSize,
						)
						.returning()
						.fetchSingle()
					val players = teamStats.getPlayers().map mapPlayer@{ playerStats ->
						return@mapPlayer context
							.insertInto(Tables.STATS_PLAYER)
							.values(
								DSL.defaultValue(Tables.STATS_PLAYER.PLAYER_ID), playerStats.nickname,
								playerStats.role.name, playerStats.totalPostLength,
								playerStats.totalPosts, playerStats.userId, teamRecord.teamId,
								playerStats.alignment.name,
							)
							.returning()
							.fetchSingle(playerMapper())

					}
					return@mapTeam mapTeam(teamRecord, players)
				}

				val actions = insertGameStats.getActions().map { actionStats ->
					return@map context
						.insertInto(Tables.STATS_ACTION)
						.values(
							DSL.defaultValue(Tables.STATS_ACTION.ACTION_ID), actionStats.actionType.name,
							actionStats.actor, actionStats.cycle, actionStats.order,
							actionStats.target, actionStats.timeStampHappened,
							actionStats.timeStampSubmitted, gameId, actionStats.phase.name,
							actionStats.additionalInfo,
						)
						.returning()
						.fetchSingle(actionMapper())
				}

				return@transactionResult GameStats(
					gameId, insertGameStats.channelId, insertGameStats.channelName,
					insertGameStats.endTime, insertGameStats.gameMode, insertGameStats.gameType,
					insertGameStats.guildId, insertGameStats.guildName, insertGameStats.startTime,
					insertGameStats.playerSize, teams.toSet(), actions.toSet(),
				)
			}

		}
	}

	fun fetchAllGameStatsOfUser(userId: Long): List<PrivacyGame> {
		return metricsService.queryTime("getAllGameStatsOfUser").time {
			jooq
				.select(
					Tables.STATS_GAME.GAME_ID,
					Tables.STATS_GAME.START_TIME,
					Tables.STATS_GAME.END_TIME,
					Tables.STATS_TEAM.ALIGNMENT,
					Tables.STATS_TEAM.IS_WINNER,
					Tables.STATS_PLAYER.NICKNAME,
					Tables.STATS_PLAYER.TOTAL_POSTS,
					Tables.STATS_PLAYER.TOTAL_POSTLENGTH,
				)
				.from(Tables.STATS_GAME)
				.join(Tables.STATS_TEAM).on(Tables.STATS_TEAM.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.join(Tables.STATS_PLAYER).on(Tables.STATS_PLAYER.TEAM_ID.eq(Tables.STATS_TEAM.TEAM_ID))
				.where(Tables.STATS_PLAYER.USER_ID.eq(userId))
				.fetch(privacyGameMapper())
		}
	}

	private fun privacyGameMapper(): RecordMapper<Record8<Long, Long, Long, String, Boolean, String, Int, Int>, PrivacyGame> {
		return RecordMapper { record ->
			PrivacyGame(
				record.get(Tables.STATS_GAME.GAME_ID),
				Instant.ofEpochMilli(record.get(Tables.STATS_GAME.START_TIME)),
				Instant.ofEpochMilli(record.get(Tables.STATS_GAME.END_TIME)),
				record.get(Tables.STATS_TEAM.ALIGNMENT),
				record.get(Tables.STATS_TEAM.IS_WINNER),
				record.get(Tables.STATS_PLAYER.NICKNAME),
				record.get(Tables.STATS_PLAYER.TOTAL_POSTS),
				record.get(Tables.STATS_PLAYER.TOTAL_POSTLENGTH), listOf(),
			)
		}
	}

	fun fetchAllActionStatsOfUser(userId: Long): Map<Long, List<PrivacyAction>> {
		return metricsService.queryTime("getAllActionStatsOfUser").time {
			jooq
				.select(
					Tables.STATS_GAME.GAME_ID,
					Tables.STATS_ACTION.ACTION_TYPE,
					Tables.STATS_ACTION.SUBMITTED,
				)
				.from(Tables.STATS_GAME)
				.join(Tables.STATS_ACTION).on(Tables.STATS_ACTION.GAME_ID.eq(Tables.STATS_GAME.GAME_ID))
				.where(Tables.STATS_ACTION.ACTOR.eq(userId))
				.fetchGroups(Tables.STATS_GAME.GAME_ID, privacyActionMapper())
		}
	}

	private fun privacyActionMapper(): RecordMapper<Record3<Long, String, Long>, PrivacyAction> {
		return RecordMapper { record ->
			PrivacyAction(
				Actions.valueOf(record.get(Tables.STATS_ACTION.ACTION_TYPE)),
				Instant.ofEpochMilli(record.get(Tables.STATS_ACTION.SUBMITTED)),
			)
		}
	}

	fun nullAllPlayerNicknamesofUser(userId: Long): Int {
		return metricsService.queryTime("nullAllPlayerNicknamesofUser").time {
			jooq.transactionResult { config ->
				config.dsl()
					.update(Tables.STATS_PLAYER)
					.set(Tables.STATS_PLAYER.NICKNAME, DSL.`val`(null, Tables.STATS_PLAYER.NICKNAME))
					.where(Tables.STATS_PLAYER.USER_ID.eq(userId))
					.execute()
			}
		}
	}
}
